src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -341,6 +381,706 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+	List *equivalenceMemberList = secondClass->equivalentAttributes;
+
+	foreach(equivalenceClassMemberCell, equivalenceMemberList)
+	{
+		AttributeEquivalenceClassMember *newEqMember = lfirst(equivalenceClassMemberCell);
+
+		if (AttributeClassContainsAttributeClassMember(newEqMember, *firstClass))
+		{
+			continue;
+		}
+
+		(*firstClass)->equivalentAttributes = lappend((*firstClass)->equivalentAttributes,
+													  newEqMember);
+	}
+}
+
+
+/*
+ * GenerateAttributeEquivalencesForJoinRestrictions gets a join restriction
+ * context and returns a list of AttrributeEquivalenceClass.
+ *
+ * The algorithm followed can be summarized as below:
+ *
+ * - Per join restriction
+ *     - Per RestrictInfo of the join restriction
+ *     - Check whether the join restriction is in the form of (Var1 = Var2)
+ *         - Create an AttributeEquivalenceClass
+ *         - Add both Var1 and Var2 to the AttributeEquivalenceClass
+ *
+ * Note that this function does not deal with whether the member of restriction
+ * are partition key or not. That's handled later in the planning.
+ */
+static List *
+GenerateAttributeEquivalencesForJoinRestrictions(JoinRestrictionContext *
+												 joinRestrictionContext)
+{
+	List *attributeEquivalenceList = NIL;
+	ListCell *joinRestrictionCell = NULL;
+
+	foreach(joinRestrictionCell, joinRestrictionContext->joinRestrictionList)
+	{
+		JoinRestriction *joinRestriction = lfirst(joinRestrictionCell);
+		ListCell *restrictionInfoList = NULL;
+
+		foreach(restrictionInfoList, joinRestriction->joinRestrictInfoList)
+		{
+			RestrictInfo *rinfo = (RestrictInfo *) lfirst(restrictionInfoList);
+			OpExpr *restrictionOpExpr = NULL;
+			Node *leftNode = NULL;
+			Node *rightNode = NULL;
+			Expr *strippedLeftExpr = NULL;
+			Expr *strippedRightExpr = NULL;
+			Var *leftVar = NULL;
+			Var *rightVar = NULL;
+			Expr *restrictionClause = rinfo->clause;
+			AttributeEquivalenceClass *attributeEquivalance = NULL;
+
+			if (!IsA(restrictionClause, OpExpr))
+			{
+				continue;
+			}
+
+			restrictionOpExpr = (OpExpr *) restrictionClause;
+			if (list_length(restrictionOpExpr->args) != 2)
+			{
+				continue;
+			}
+			if (!OperatorImplementsEquality(restrictionOpExpr->opno))
+			{
+				continue;
+			}
+
+			leftNode = linitial(restrictionOpExpr->args);
+			rightNode = lsecond(restrictionOpExpr->args);
+
+			/* we also don't want implicit coercions */
+			strippedLeftExpr = (Expr *) strip_implicit_coercions((Node *) leftNode);
+			strippedRightExpr = (Expr *) strip_implicit_coercions((Node *) rightNode);
+
+			if (!(IsA(strippedLeftExpr, Var) && IsA(strippedRightExpr, Var)))
+			{
+				continue;
+			}
+
+			leftVar = (Var *) strippedLeftExpr;
+			rightVar = (Var *) strippedRightExpr;
+
+			attributeEquivalance = palloc0(sizeof(AttributeEquivalenceClass));
+			attributeEquivalance->equivalenceId = attributeEquivalenceId++;
+
+			/* we could (probably) safely do that since we disallowed ANTI JOINs ? */
```

Comment:
This is an old comment and removed now. We currently do not disallow anti joins. I don't see a reason to disallow anti-joins. Do you see?